src/typed_ast.cpp

```diff
@@ -4,222 +4,225 @@
 
 namespace TypedAST {
 
-Own<TypedAST> convert_ast(AST::IntegerLiteral* ast) {
-	auto typed_integer = std::make_unique<IntegerLiteral>();
+TypedAST* convert_ast(AST::IntegerLiteral* ast, Allocator& alloc) {
+	auto typed_integer = alloc.make<IntegerLiteral>();
 	typed_integer->m_token = ast->m_token;
 	return typed_integer;
 }
 
-Own<TypedAST> convert_ast(AST::NumberLiteral* ast) {
-	auto typed_number = std::make_unique<NumberLiteral>();
+TypedAST* convert_ast(AST::NumberLiteral* ast, Allocator& alloc) {
+	auto typed_number = alloc.make<NumberLiteral>();
 	typed_number->m_token = ast->m_token;
 	return typed_number;
 }
 
-Own<TypedAST> convert_ast(AST::StringLiteral* ast) {
-	auto typed_string = std::make_unique<StringLiteral>();
+TypedAST* convert_ast(AST::StringLiteral* ast, Allocator& alloc) {
+	auto typed_string = alloc.make<StringLiteral>();
 	typed_string->m_token = ast->m_token;
 	return typed_string;
 }
 
-Own<TypedAST> convert_ast(AST::BooleanLiteral* ast) {
-	auto typed_boolean = std::make_unique<BooleanLiteral>();
+TypedAST* convert_ast(AST::BooleanLiteral* ast, Allocator& alloc) {
+	auto typed_boolean = alloc.make<BooleanLiteral>();
 	typed_boolean->m_token = ast->m_token;
 	return typed_boolean;
 }
 
-Own<TypedAST> convert_ast(AST::NullLiteral* ast) {
-	return std::make_unique<NullLiteral>();
+TypedAST* convert_ast(AST::NullLiteral* ast, Allocator& alloc) {
+	return alloc.make<NullLiteral>();
 }
 
-Own<TypedAST> convert_ast(AST::ArrayLiteral* ast) {
-	auto typed_array = std::make_unique<ArrayLiteral>();
+TypedAST* convert_ast(AST::ArrayLiteral* ast, Allocator& alloc) {
+	auto typed_array = alloc.make<ArrayLiteral>();
 
 	for (auto element : ast->m_elements) {
-		typed_array->m_elements.push_back(convert_ast(element));
+		typed_array->m_elements.push_back(convert_ast(element, alloc));
 	}
 
 	return typed_array;
 }
 
-Own<TypedAST> convert_ast(AST::DictionaryLiteral* ast) {
-	auto typed_dict = std::make_unique<DictionaryLiteral>();
+TypedAST* convert_ast(AST::DictionaryLiteral* ast, Allocator& alloc) {
+	auto typed_dict = alloc.make<DictionaryLiteral>();
 
 	for (auto element : ast->m_body) {
-		typed_dict->m_body.push_back(convert_ast(element));
+		typed_dict->m_body.push_back(convert_ast(element, alloc));
 	}
 
 	return typed_dict;
 }
 
-Own<TypedAST> convert_ast(AST::FunctionLiteral* ast) {
-	auto typed_function = std::make_unique<FunctionLiteral>();
+TypedAST* convert_ast(AST::FunctionLiteral* ast, Allocator& alloc) {
+	auto typed_function = alloc.make<FunctionLiteral>();
 
 	for (auto arg : ast->m_args) {
 		assert(arg->type() == ASTTag::Declaration);
-		auto* decl = static_cast<AST::Declaration*>(arg);
+		auto decl = static_cast<AST::Declaration*>(arg);
 
-		Declaration typed_decl;
-		typed_decl.m_identifier_token = decl->m_identifier_token;
-		typed_decl.m_surrounding_function = typed_function.get();
-		typed_function->m_args.push_back(std::move(typed_decl));
+		auto typed_decl = alloc.make<Declaration>();
```

Comment:
Why did you change this one from a value to a pointer?

Reply:
This one, honestly, it was a matter of consistency to have them all in the allocator... I guess there are many places where that could be avoided just by not allocating them in the node allocator. I should rethink that, I also did it in the last pr.

Reply:
As a rule of thumb, the less indirection the better :)